core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala

```diff
@@ -208,6 +201,87 @@ class CompactEntityQuery(driverComponent: DriverComponent) extends RawSqlQuery w
+
+    allCopies.map { copyActionResults: Iterator[Int] =>
+      copyActionResults.sum
+    }
+  }
+
+  def getEntitySubtrees(workspaceId: UUID,
+                        entityType: String,
+                        entityNames: Set[String]
+                       ): ReadAction[Seq[EntityPath]] = {
+    val refs = entityNames.map { name => AttributeEntityReference(entityType, name) }
+    for {
+      startingEntityRecords <- getEntityRefs(workspaceId, refs)
+      entities = startingEntityRecords.map { record => record.toAttributeEntityReference}
+      refsToId = startingEntityRecords.map(record => EntityPath(Seq(record.toAttributeEntityReference)) -> record.toAttributeEntityReference).toMap
+      allRefs <- recursiveGetEntityReferences(workspaceId, entities.toSet, refsToId)
+    } yield allRefs
+  }
+
+  def recursiveGetEntityReferences(workspaceId: UUID,
+                                   entities: Set[AttributeEntityReference],
+                                   accumulatedPathsWithLastId: Map[EntityPath, AttributeEntityReference]
+                                  ): ReadAction[Seq[EntityPath]] = {
+    val actions: Seq[ReadAction[Seq[(AttributeEntityReference, AttributeEntityReference)]]] = entities.toSeq.map { entity =>
+      getReferencesTo(workspaceId, Seq(entity)).map { references =>
+        references.map(ref => (entity, ref))
+      }
+    }
+
+    DBIO.sequence(actions).map(_.flatten.toSet).flatMap { priorEntityWithCurrentRec =>
+      val currentPaths = priorEntityWithCurrentRec.flatMap { case (priorEntity, currentRec) =>
+        val pathsThatEndWithPrior = accumulatedPathsWithLastId.filter { case (_, entity) => entity == priorEntity }
+        pathsThatEndWithPrior.keys.map(_.path).map { path =>
+          (EntityPath(path :+ currentRec), currentRec)
+        }
+      }.toMap
+
+      val untraversedIds = priorEntityWithCurrentRec.map(_._2) -- accumulatedPathsWithLastId.values.toSet
+      if (untraversedIds.isEmpty) {
+        DBIO.successful(accumulatedPathsWithLastId.keys.toSeq)
+      } else {
+        recursiveGetEntityReferences(workspaceId, untraversedIds, accumulatedPathsWithLastId ++ currentPaths)
+      }
+    }
+  }
+
+  def copyEntities(sourceWorkspaceId: UUID, destWorkspaceId: UUID, refs: Set[AttributeEntityReference]): ReadWriteAction[Int] =
+    if (refs.isEmpty) {
+      DBIO.successful(0)
+    } else {
+      val typeNameClauses = generateTypeNameSql(refs)
+      val sql = concatSqlActions(
+        sql"""insert into ENTITY(name, entity_type, workspace_id, record_version, deleted, attributes)
+             select name, entity_type, $destWorkspaceId, record_version, 0, attributes
+             from ENTITY e
+             where e.workspace_id = $sourceWorkspaceId
+             and deleted = 0
+             and ( """,
+        reduceSqlActionsWithDelim(typeNameClauses.toSeq, sql" or "))
+             sql""" );"""
+      sql.asUpdate
+  }
```

Comment:
this looks correct for the `ENTITY` table. You'll need a second `insert into … select` statement to populate the `ENTITY_REFS` table for the `destWorkspaceId`, too.